packages/playwright/src/worker/workerMain.ts

```diff
@@ -186,28 +193,43 @@ export class WorkerMain extends ProcessRunner {
       void this._stop();
   }
 
-  private async _loadIfNeeded() {
+  private async _load(): Promise<{
+    config: FullConfigInternal,
+    project: FullProjectInternal,
+    poolBuilder: PoolBuilder,
+  } | undefined> {
     if (this._config)
-      return;
+      return { config: this._config, project: this._project, poolBuilder: this._poolBuilder };
 
     this._config = await deserializeConfig(this._params.config);
-    this._project = this._config.projects.find(p => p.id === this._params.projectId)!;
+    const project = this._config.projects.find(p => p.id === this._params.projectId);
+    if (!project)
+      return undefined;
+    this._project = project;
     this._poolBuilder = PoolBuilder.createForWorker(this._project);
+
+    return { config: this._config, project: this._project, poolBuilder: this._poolBuilder };
   }
 
   async runTestGroup(runPayload: RunPayload) {
     this._runFinished = new ManualPromise<void>();
     const entries = new Map(runPayload.entries.map(e => [e.testId, e]));
-    let fatalUnknownTestIds;
+    let fatalUnknownTestIds: string[] | undefined;
     try {
-      await this._loadIfNeeded();
-      const fileSuite = await loadTestFile(runPayload.file, this._config.config.rootDir);
-      const suite = bindFileSuiteToProject(this._project, fileSuite);
+      const workerState = await this._load();
```

Comment:
I do not understand why do we have to use the return value now, instead of reading from `this._config` and similar. Perhaps just return a boolean, or better yet set `this._isMissingProject` inside `_loadIfNeeded()` and check it here?

Reply:
I considered that to be bad control flow, since clearly you could continue executing without checking `this._isMissingProject`. I'll change it though.